### PR TITLE
[Metal] Add support for xcode 13 beta5 .

### PR DESCRIPTION
**Pull request by mandel-macaque**

### CODE REVIEW

**Review by chamons — APPROVED** [review]

File: src/Metal/MTLDevice.cs [hunk 1]
```diff
@@ -50,13 +50,21 @@ public static partial class MTLDevice {
 				return system_default;
 			}
 		}
+		
+#if MONOMAC || __MACCATALYST__
```
Comment (rolfbjarne) [inline src/Metal/MTLDevice.cs:54]: This should be run on Mac Catalyst as well now:

https://github.com/xamarin/xamarin-macios/blob/9286bfb79309811b96ece9a283f1d13d5f7c52e3/tests/monotouch-test/Metal/MTLDeviceTests.cs#L24-L38

File: tests/introspection/iOS/iOSApiSelectorTest.cs [hunk 2]
```diff
@@ -106,7 +106,12 @@ protected override bool Skip (Type type)
 			// iOS 10 - this type can only be instantiated on devices, but the selectors are forwarded
 			//  to a MTLHeapDescriptorInternal and don't respond - so we'll add unit tests for them
 			case "MTLHeapDescriptor":
-				return Runtime.Arch == Arch.DEVICE;
+			case "MTLLinkedFunctions":
+			case "MTLRenderPipelineDescriptor":
+			case "MTLTileRenderPipelineDescriptor":
+			case "MTLRasterizationRateLayerDescriptor":
+			case "MTLRasterizationRateMapDescriptor":
+				return Runtime.Arch == Arch.SIMULATOR;
```
Comment (rolfbjarne) [inline tests/introspection/iOS/iOSApiSelectorTest.cs:114]: Why did this change from `== Arch.DEVICE` to `== Arch.SIMULATOR`? Was that intentional?

File: src/metal.cs [hunk 3]
```diff
@@ -4397,25 +4639,32 @@ interface MTLRasterizationRateLayerDescriptor : NSCopying {
 		[Internal]
 		[Export ("initWithSampleCount:horizontal:vertical:")]
 		IntPtr Constructor (MTLSize sampleCount, IntPtr horizontal, IntPtr vertical);
-
-/* Selectors reported as not working by introspection: https://github.com/xamarin/maccore/issues/1976
```
Comment (rolfbjarne) [inline src/metal.cs:4401]: This code should be enabled too:

https://github.com/xamarin/xamarin-macios/blob/9286bfb79309811b96ece9a283f1d13d5f7c52e3/src/Metal/MTLRasterizationRateLayerDescriptor.cs#L15-L35

File: src/metal.cs [hunk 4]
```diff
@@ -5243,4 +5563,151 @@ interface MTLFunctionLogDebugLocation {
...
+		IMTLFunction[] TileAdditionalBinaryFunctions { get; set; }
+	}
+
+	[Mac (12, 0), iOS (15, 0), NoTV, MacCatalyst (15, 0), NoWatch]
+	[BaseType (typeof (NSObject))]
+	interface MTLMotionKeyframeData {
+		[NullAllowed, Export ("buffer", ArgumentSemantic.Retain)]
+		IMTLBuffer Buffer { get; set; }
+
+		[Export ("offset")] nuint Offset { get; set; }
```
Comment (rolfbjarne) [inline src/metal.cs:5597]:
```suggestion
		[Export ("offset")]
		nuint Offset { get; set; }
```

File: src/metal.cs [hunk 5]
```diff
@@ -5243,4 +5563,151 @@ interface MTLFunctionLogDebugLocation {
...
+
+		[Static]
+		[Export ("data")]
+		MTLMotionKeyframeData Data ();
```
Comment (rolfbjarne) [inline src/metal.cs:5601]: Needs a verb:

```suggestion
		MTLMotionKeyframeData Create ();
```

File: src/metal.cs [hunk 6]
```diff
@@ -5243,4 +5563,151 @@ interface MTLFunctionLogDebugLocation {
...
+
+		[Export ("indexType", ArgumentSemantic.Assign)]
+		MTLIndexType IndexType { get; set; }
+
+		[Export ("triangleCount")]
+		nuint TriangleCount { get; set; }
+
+		[Static]
+		[Export ("descriptor")]
+		MTLAccelerationStructureMotionTriangleGeometryDescriptor Descriptor ();
```
Comment (rolfbjarne) [inline src/metal.cs:5693]: Needs a verb:

```suggestion
		MTLAccelerationStructureMotionTriangleGeometryDescriptor Create ();
```

File: src/metal.cs [hunk 7]
```diff
@@ -5243,4 +5563,151 @@ interface MTLFunctionLogDebugLocation {
...
+
+		[Export ("boundingBoxStride")]
+		nuint BoundingBoxStride { get; set; }
+
+		[Export ("boundingBoxCount")]
+		nuint BoundingBoxCount { get; set; }
+
+		[Static]
+		[Export ("descriptor")]
+		MTLAccelerationStructureMotionBoundingBoxGeometryDescriptor Descriptor ();
```
Comment (rolfbjarne) [inline src/metal.cs:5711]: Same:

```suggestion
		MTLAccelerationStructureMotionBoundingBoxGeometryDescriptor Create ();
```

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: tests/introspection/iOS/iOSApiSelectorTest.cs (hunk 2 above)
Comment (mandel-macaque) [reply tests/introspection/iOS/iOSApiSelectorTest.cs:114]: Yes. The skip should be when we are on simulator, not the other way around. I guess it was a mistake in a previous change

File: src/Metal/MTLDevice.cs (hunk 1 above)
Comment (mandel-macaque) [reply src/Metal/MTLDevice.cs:54]: nice catch.

File: src/Metal/MTLDevice.cs [hunk 8]
```diff
@@ -50,13 +50,21 @@ public static partial class MTLDevice {
...
 
-#if MONOMAC
+#if !NET
```
Comment (tj-devel709) [inline src/Metal/MTLDevice.cs:56]: Are these two methods no longer supported in Mac?

Comment (tj-devel709) [reply src/Metal/MTLDevice.cs:56]: Or are they just getting the default now?

File: src/metal.cs [hunk 9]
```diff
@@ -157,10 +157,11 @@ partial interface MTLBuffer : MTLResource {
 
 		[Abstract, Export ("contents")]
 		IntPtr Contents { get; }
-#if MONOMAC
+	
+		[NoiOS, NoTV, MacCatalyst (15,0)]
 		[Abstract, Export ("didModifyRange:")]
 		void DidModify (NSRange range);
-#endif
+
```
Comment (tj-devel709) [inline src/metal.cs:164]: nit: extra tab

File: src/metal.cs [hunk 10]
```diff
@@ -4397,25 +4639,32 @@ interface MTLRasterizationRateLayerDescriptor : NSCopying {
...
+		[MacCatalyst (15,0)]
 		[Export ("horizontal")]
 		MTLRasterizationRateSampleArray Horizontal { get; }
 
+		[MacCatalyst (15,0)]
 		[Export ("vertical")]
 		MTLRasterizationRateSampleArray Vertical { get; }
- */
+ 
+		[Mac (12, 0), iOS (15, 0), TV (15,0), MacCatalyst(15,0), NoWatch]
```
Comment (tj-devel709) [inline src/metal.cs:4661]:
```suggestion
		[Mac (12, 0), iOS (15, 0), TV (15,0), MacCatalyst (15,0), NoWatch]
```
nit: spacing

File: src/metal.cs [hunk 11]
```diff
@@ -4397,25 +4639,32 @@ interface MTLRasterizationRateLayerDescriptor : NSCopying {
...
+		[Export ("maxSampleCount")]
+		MTLSize MaxSampleCount { get; }
+
+		[Mac (12, 0), iOS (15, 0), TV (15,0), MacCatalyst(15,0), NoWatch]
```
Comment (tj-devel709) [inline src/metal.cs:4665]:
```suggestion
		[Mac (12, 0), iOS (15, 0), TV (15,0), MacCatalyst (15,0), NoWatch]
```
nit: spacing

File: src/metal.cs [hunk 12]
```diff
@@ -5243,4 +5563,152 @@ interface MTLFunctionLogDebugLocation {
...
+		[NullAllowed, Export ("fragmentAdditionalBinaryFunctions", ArgumentSemantic.Copy)]
+		IMTLFunction[] FragmentAdditionalBinaryFunctions { get; set; }
+
+		[NullAllowed, Export ("tileAdditionalBinaryFunctions", ArgumentSemantic.Copy)]
+		IMTLFunction[] TileAdditionalBinaryFunctions { get; set; }
+	}
+
+	[Mac (12, 0), iOS (15, 0), NoTV, MacCatalyst (15, 0), NoWatch]
+	[BaseType (typeof (NSObject))]
+	interface MTLMotionKeyframeData {
```
Comment (tj-devel709) [inline src/metal.cs:5593]:
```suggestion
	interface MtlMotionKeyframeData {
```
Not sure, but wanted to check this?

File: src/metal.cs [hunk 13]
```diff
@@ -5243,4 +5563,152 @@ interface MTLFunctionLogDebugLocation {
 		nuint Column { get; }
 	}
 
+	[Mac (12,0), iOS (15,0), TV (15,0), MacCatalyst (15, 0), NoWatch]
+	[BaseType (typeof (NSObject))]
+	interface MTLStitchedLibraryDescriptor : NSCopying
```
Comment (tj-devel709) [inline src/metal.cs:5568]: Should all these "MTL..." classes be "Mtl..." ?

**Review by rolfbjarne — APPROVED** [review]

**Pull request merged**